### PR TITLE
Create an authorized sale

### PR DESCRIPTION
**Pull request by antoniofilho**
This PR introduces:

* Models structure
* Sales Authorization
* Configurations and gem setup

### CODE REVIEW

File: lib/braspag-rest/request.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+module BraspagRest
+  class Request
+    class << self
+      SALE_ENDPOINT = '/v2/sales/'
+
+      def authorize(request_id, params)
+        gateway_response = RestClient.post(sale_url, params.to_json, default_headers.merge('RequestId' => request_id))
+        BraspagRest::Response.new(gateway_response)
+      rescue RestClient::ExceptionWithResponse => e
+        config.logger.warn("[BraspagRest] #{e} #slack") if config.log_enabled?
```
Comment (pmatiello) [inline lib/braspag-rest/request.rb]: Since this is a public project, the "#slack" tag might not really make much sense.

Comment (antoniofilho) [inline lib/braspag-rest/request.rb]: is true... I forgot :smile:

File: lib/braspag-rest/sale.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    coerce_key :customer, BraspagRest::Customer
+    coerce_key :payment, BraspagRest::Payment
+
+    def save
+      response = BraspagRest::Request.authorize(request_id, inverse_attributes)
+
+      if response.success?
+        initialize_attributes(response.parsed_body)
+      else
+        initialize_errors(response.parsed_body) and return false
```
Comment (pmatiello) [inline lib/braspag-rest/sale.rb]: Cute. :)

**Comment by pmatiello** [comment]
:credit_card: :money_with_wings:

File: lib/braspag-rest/hashie/iutrash.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+module Hashie
+  class IUTrash < Hashie::IUDash
+    include Hashie::Extensions::Dash::PropertyTranslation
+
+    def inverse_attributes
+      {}.tap do |attributes|
+        self.class.translations.each do |(from, property)|
```
Comment (bvicenzo) [inline lib/braspag-rest/hashie/iutrash.rb]: Are these parenthesis realy necessary?

Comment (antoniofilho) [inline lib/braspag-rest/hashie/iutrash.rb]: No, before I'm using translations_hash that was a little bit different and needs that, but this translations is a simple hash... fixed 933c4dbf95b3bce3b8fc555bb578f01dc43a96c9

File: lib/braspag-rest/request.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
+module BraspagRest
+  class Request
+    class << self
+      SALE_ENDPOINT = '/v2/sales/'
+
+      def authorize(request_id, params)
+        gateway_response = RestClient.post(sale_url, params.to_json, default_headers.merge('RequestId' => request_id))
+        BraspagRest::Response.new(gateway_response)
+      rescue RestClient::ExceptionWithResponse => e
+        config.logger.warn("[BraspagRest] #{e}") if config.log_enabled?
```
Comment (bvicenzo) [inline lib/braspag-rest/request.rb]: I guess could be good to put at least the first backtrace line here:
```config.logger.warn("[BraspagRest] #{e} - #{e.backtrace.first}") if config.log_enabled?```

File: lib/braspag-rest/request.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        BraspagRest::Response.new(e.response)
+      rescue RestClient::Exception => e
+        config.logger.error("[BraspagRest] #{e}") if config.log_enabled?
```
Comment (bvicenzo) [inline lib/braspag-rest/request.rb]: and here...

**Comment by bvicenzo** [comment]
:credit_card: :flower_playing_cards:

File: lib/braspag-rest/request.rb (hunk 4 above)
Comment (antoniofilho) [inline lib/braspag-rest/request.rb]: I don't know if this is really necessary, because this exception and the following are only raised by line 7.

Comment (bvicenzo) [inline lib/braspag-rest/request.rb]: :+1:

**Comment by milare** [comment]
:+1: :credit_card:

**Pull request merged**